### PR TITLE
PS-5006: Port storage engine enforcement to 8.0

### PR DESCRIPTION
**Pull request by inikep**
1. The patch is based on https://github.com/percona/percona-server/commit/7ad12e4df3880ac11fbf7656a14bf9b8043d25e1 with improvements at https://github.com/percona/percona-server/commit/868dbf4ba27ac4974b98f11c835ae367bdeb0022 and https://github.com/percona/percona-server/commit/1b93144cd839fcb804e6c0fbe8bb4a30de27dce5
2. Added exceptions for transactional data dictionary (DD) tables (enforcement is forbidden)
3. `percona_enforce_se_create_table_like` was changed as `mysql.proc` was removed in 8.0
4. Test `ALTER TABLE with ALGORITHM=INSTANT,INPLACE,COPY` in percona_enforce_storage_engine_alter_table.inc

### CODE REVIEW

File: mysql-test/include/percona_enforce_storage_engine_alter_table.inc [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# OPTIMIZE TABLE must not change table engine
+# This statement will use "ALGORITHM=INPLACE" in both 5.6 and 5.7
```
Comment (laurynas-biveinis) [inline mysql-test/include/percona_enforce_storage_engine_alter_table.inc]: Does "AlGORITHM=INPLACE" hold in 8.0 too? Please edit comment

File: mysql-test/include/percona_enforce_storage_engine_alter_table.inc [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+--echo After OPTIMIZE TABLE statement
+# Make sure table engine hasn't been changed
+# Make sure column definition hasn't been changed
+--let $enforce_engine= InnoDB
+--let $enforce_column= varchar(1)
+--source include/percona_enforce_storage_engine_alter_table_assert.inc
+
+# Modify table definition (do not specify "ENGINE")
+# This statement will use "ALGORITHM=COPY" in both 5.6 and 5.7
```
Comment (laurynas-biveinis) [inline mysql-test/include/percona_enforce_storage_engine_alter_table.inc]: Does "AlGORITHM=COPY" hold in 8.0 too? Please edit comment

File: mysql-test/t/percona_enforce_storage_engine.test [hunk 3]
```diff
@@ -0,0 +1,27 @@
+--source include/have_myisam.inc
+
+call mtr.add_suppression("is not the same as enforced storage engine");
+
+--disable_warnings
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_enforce_storage_engine.test]: Drop obsolete idiom

File: sql/handler.h [hunk 4]
```diff
@@ -4169,7 +4169,7 @@ class handler {
 
   virtual bool is_ignorable_error(int error);
   MY_NODISCARD virtual bool continue_partition_copying_on_error(
-    int error MY_ATTRIBUTE((unused))) {
+      int error MY_ATTRIBUTE((unused))) {
```
Comment (laurynas-biveinis) [inline sql/handler.h:4172]: ?

File: sql/mysqld.cc [hunk 5]
```diff
@@ -5466,6 +5468,43 @@ static int init_server_components() {
              "default_tmp_storage_engine", default_tmp_storage_engine);
   }
 
+  /*
+    Validate any enforced storage engine
+  */
+  if (enforce_storage_engine && !opt_initialize && !opt_noacl) {
+    LEX_STRING name = {enforce_storage_engine, strlen(enforce_storage_engine)};
+    plugin_ref plugin;
+    if ((plugin = ha_resolve_by_name(0, &name, false))) {
```
Comment (laurynas-biveinis) [inline sql/mysqld.cc]: s/0/nullptr

File: sql/mysqld.cc [hunk 6]
```diff
@@ -5466,6 +5468,43 @@ static int init_server_components() {
...
+      handlerton *hton = plugin_data<handlerton *>(plugin);
+      LEX_STRING defname = {default_storage_engine,
+                            strlen(default_storage_engine)};
+      plugin_ref defplugin;
+      handlerton *defhton;
+      if ((defplugin = ha_resolve_by_name(0, &defname, false))) {
```
Comment (laurynas-biveinis) [inline sql/mysqld.cc]: s/0/nullptr

File: sql/mysqld.cc [hunk 7]
```diff
@@ -5466,6 +5468,43 @@ static int init_server_components() {
...
+        }
+      }
+      if (ha_is_storage_engine_disabled(hton)) {
+        sql_print_error(
+            "enforced storage engine %s is among disabled storage "
+            "engines",
+            enforce_storage_engine);
+        unireg_abort(MYSQLD_ABORT_EXIT);
+      }
+      plugin_unlock(0, defplugin);
```
Comment (laurynas-biveinis) [inline sql/mysqld.cc]: s/0/nullptr

File: sql/mysqld.cc [hunk 8]
```diff
@@ -5466,6 +5468,43 @@ static int init_server_components() {
...
+      plugin_unlock(0, plugin);
```
Comment (laurynas-biveinis) [inline sql/mysqld.cc]: s/0/nullptr

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
Minor comments only

**Comment by inikep** [comment]
All comments were addressed.
Moreover I added tests for `ALTER TABLE` with `ALGORITHM=INSTANT,INPLACE,COPY` in `percona_enforce_storage_engine_alter_table.inc`.

File: sql/mysqld.cc [hunk 9]
```diff
@@ -4837,7 +4839,7 @@ static bool initialize_storage_engine(char *se_name, const char *se_kind,
   LEX_STRING name = {se_name, strlen(se_name)};
   plugin_ref plugin;
   handlerton *hton;
-  if ((plugin = ha_resolve_by_name(0, &name, false)))
+  if ((plugin = ha_resolve_by_name(nullptr, &name, false)))
```
Comment (laurynas-biveinis) [inline sql/mysqld.cc]: Please do not cleanup otherwise untouched code

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
OK to push with the comment addressed

**Pull request merged**